lib/coffee_place/importer.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require 'csv'
+require 'open-uri'
+
+require_relative './result'
+require_relative './location'
+
+module CoffeePlace
+  # Imports locations from local or remote CSV file.
+  class Importer
+    def initialize
+      @locations = []
+      @errors = []
+    end
+
+    def import_from(source_name)
+      line_index = 0
+
+      with_source(source_name) do |csv|
+        CSV.parse(csv) do |line|
+          line_index += 1
+          import_line(line, line_index)
+        end
+      end
+
+      location_results
+    end
+
+    def import_line(line, line_index)
+      place_name, lat, lon = line
+
+      result = Location.validate(name: place_name, lat: lat, lon: lon)
+
+      if result.success?
+        @locations << result.value
+      else
+        add_import_error "CSV error on line #{line_index}: #{result.error}"
+      end
+    end
+
+    def location_results
+      if @errors.empty?
+        Result.success(@locations)
+      else
+        Result.failure(@errors)
+      end
+    end
+
+    def with_source(source_name, &block)
+      uri = URI.parse(source_name)
+
+      # When we have an URL, download remote file
+      return with_remote_file(uri, &block) if uri.scheme
+
+      # When we have a local file, try to open it
+      return with_local_file(uri, &block) if File.exist?(uri.path)
+
+      add_import_error unknown_source_error_message(source_name)
+    rescue URI::InvalidURIError
+      add_import_error "#{source_name} is not a valid URL or local filename"
+    end
+
+    def with_remote_file(uri, &block)
+      uri.open(&block)
+    rescue OpenURI::HTTPError => e
+      add_import_error "Failed to download remote file \"#{uri}\": #{e.message}"
+    end
+
+    def with_local_file(uri, &block)
+      File.open(uri.path, &block)
+    rescue StandardError => e
+      add_import_error "Failed to read local file \"#{uri.path}\": #{e.message}"
+    end
+
+    def unknown_source_error_message(source_name)
+      <<~ERROR
+        Don't know how to import locations from: #{source_name.inspect}
+        Make sure the URL or filename is not misstyped.
+      ERROR
```

Comment:
How would you proceed if we'd want to re-use this error message in 10 other places?

Reply:
I tried to break the short novel I wrote in the previous comment, into multiple logical chapters.
Stay tuned for it's sequel!

Reply:
My approach is usually to just extract them to a `locales/errors/en.yml` or something similar and just rely on `I18n.t`. Worked like a charm so far, you can do interpolation if needed too.